tests/strategy/test_strategy_helpers.py

```diff
@@ -95,3 +97,38 @@ def test_merge_informative_pair_lower():
 
     with pytest.raises(ValueError, match=r"Tried to merge a faster timeframe .*"):
         merge_informative_pair(data, informative, '1h', '15m', ffill=True)
+
+
+def test_stoploss_from_open():
+    open_price_ranges = [
+        [0.01, 1.00, 30],
+        [1, 100, 30],
+        [100, 10000, 30],
+    ]
+    current_profit_range = [-0.99, 2, 30]
+    desired_stop_range = [-0.50, 0.50, 30]
+
+    for open_range in open_price_ranges:
+        for open_price in np.linspace(*open_range):
+            for desired_stop in np.linspace(*desired_stop_range):
+
+                # -1 is not a valid current_profit, should return 1
+                assert stoploss_from_open(desired_stop, -1) == 1
+
+                for current_profit in np.linspace(*current_profit_range):
+                    current_price = open_price * (1 + current_profit)
+                    expected_stop_price = open_price * (1 + desired_stop)
+
+                    stoploss = stoploss_from_open(desired_stop, current_profit)
+
+                    assert stoploss >= 0
+                    assert stoploss <= 1
+
+                    stop_price = current_price * (1 - stoploss)
+
+                    # there is no correct answer if the expected stop price is above
+                    # the current price
+                    if expected_stop_price > current_price:
+                        assert stoploss == 0
```

Comment:
just running my thoughts past you more than anything...

If i call this with `stoploss_from_open(-0.5, -0.57)` ... i get 0.0. 
i guess this is expected, as obviously, current-profit is lower than the "desired" stoploss location. 

By returning 0.0 in this case, custom-stoploss will not be adjusted.

Reply:
Right, I think this is mostly a question of what you would like the behaviour to be in this case.  

I added this rule because while it's mathematically possible to return a result that is higher than the current price, it doesn't make any sense to return that value from custom_stoploss(), especially since the abs() value will be used, giving a totally wrong result.

I believe returning 0 from custom_stoploss() will cause the bot to log a warning, which I think is a good compromise behaviour.  It will tell you something is wrong, but not fatally so.

Another possibility would be to return a chosen very small value such as 0.0001, to cause the stoploss to be set nearly at the current price.  Since the user asked for a stoploss above the current price, this would be the closest possible result that is still valid.  This is actually the behaviour I would personally want most of the time, but I can also easily achieve it by returning max(0.0001, result), if returning a small value seems too arbitrary for default behaviour.

Reply:
yes, it's mostly a question - and causing the warning to be logged sounds good to me.

an alternative might be to return a high number (1?) - causing the stop to be ignored (without log spamming) - but it'll lead to users not noticing this behavior - and usually it'll point to a faulty logic in the stoploss function - so i guess this will be ok as is.


you'll need to be careful with values too close to current price - if price moves against you in that moment (price always moves), you'll cause a emergency sell (if the current price dropped below the stop price) - at least if you're using stoploss on exchange.